jmespath-core/src/main/java/io/burt/jmespath/function/MergeFunction.java

```diff
@@ -0,0 +1,43 @@
+package io.burt.jmespath.function;
+
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Map;
+import java.util.LinkedHashMap;
+
+import io.burt.jmespath.Adapter;
+
+public class MergeFunction extends JmesPathFunction {
+  public MergeFunction() {
+    super(1, Integer.MAX_VALUE);
+  }
+
+  @Override
+  protected <T> T internalCall(Adapter<T> adapter, List<ExpressionOrValue<T>> arguments) {
+    if (isObjectArray(adapter, arguments)) {
+      Map<String, T> accumulator = new LinkedHashMap<>();
+      for (ExpressionOrValue<T> argument : arguments) {
+        T value = argument.value();
+        for (String property : adapter.getPropertyNames(value)) {
+          accumulator.put(property, adapter.getProperty(value, property));
+        }
```

Comment:
From an adapter perspective, I would prefer if we didn't have to go through intermediate Java objects. In this case, it would be nicer to iterate over the entries of the objects, and have a method for creating a new JSON object incrementally. But I guess given the previous discussion that such modifications of the adaptor interface will come later (if at all).